util/grpc/src/auth/token_authenticator.rs

```diff
@@ -0,0 +1,217 @@
+        let signature = parts
+            .next()
+            .ok_or(AuthenticatorError::InvalidAuthorizationToken)?;
+        if parts.next().is_some() {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        if username != credentials.username {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        if !self.is_valid_time(timestamp, SystemTime::now())? {
+            return Err(AuthenticatorError::ExpiredAuthorizationToken);
+        }
+        if !self.is_valid_signature(&format!("{}:{}", username, timestamp), signature)? {
+            return Err(AuthenticatorError::InvalidAuthorizationToken);
+        }
+        Ok(credentials.username)
+    }
+}
+
+impl TokenAuthenticator {
+    pub fn new(shared_secret: [u8; 32], max_token_lifetime: Duration) -> Self {
+        Self {
+            shared_secret,
+            max_token_lifetime,
+        }
+    }
+
+    fn is_valid_time(&self, timestamp: &str, now: SystemTime) -> Result<bool, AuthenticatorError> {
+        let token_time: Duration = Duration::from_secs(
+            timestamp
+                .parse()
+                .map_err(|_| AuthenticatorError::InvalidAuthorizationToken)?,
+        );
+        let our_time: Duration = now
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .map_err(|_| AuthenticatorError::ExpiredAuthorizationToken)?;
+        let distance: Duration = our_time
+            .checked_sub(token_time)
+            .unwrap_or_else(|| token_time - our_time);
+        Ok(distance < self.max_token_lifetime)
+    }
+
+    fn is_valid_signature(&self, data: &str, signature: &str) -> Result<bool, AuthenticatorError> {
+        let mut mac = Hmac::<sha2::Sha256>::new_varkey(&self.shared_secret)
+            .map_err(|_| AuthenticatorError::Other("Invalid HMAC key".to_owned()))?;
+        mac.update(data.as_bytes());
+        let our_signature = mac.finalize().into_bytes();
+
+        let their_suffix: Vec<u8> =
+            hex::decode(signature).map_err(|_| AuthenticatorError::InvalidAuthorizationToken)?;
+        let our_suffix: &[u8] = &our_signature[..10];
+        Ok(bool::from(our_suffix.ct_eq(&their_suffix)))
+    }
+}
+
+/// Error values for token generator.
+#[derive(Display, Debug)]
+pub enum TokenBasicCredentialsGeneratorError {
+    /// SystemTime error
+    SystemTime,
+
+    /// Invalid HMAC key
+    InvalidHmacKey,
+}
+
+/// Token generator - an object that can generate HMAC authentication tokens.
+#[derive(Zeroize)]
+pub struct TokenBasicCredentialsGenerator {
+    shared_secret: [u8; 32],
+}
+
+impl TokenBasicCredentialsGenerator {
+    pub fn new(shared_secret: [u8; 32]) -> Self {
+        Self { shared_secret }
+    }
+
+    pub fn generate_for(
+        &self,
+        user_id: &str,
+    ) -> Result<BasicCredentials, TokenBasicCredentialsGeneratorError> {
+        let current_time_seconds = SystemTime::now()
```

Comment:
This shows up in various places, and it's probably OK, but one thing to consider is adding a `TimeProvider::from_epoch()` trait / method, and implementing it in terms of `SystemTime`, then making these things all generic over that (e.g. `TokenThing<TimeProvider = SystemTime>`).

At minimum, this will allow mocking out the source of "now" to make all tests deterministic, as well as let us test weird values and bounds conditions.